internal/admin/admin.go

```diff
@@ -197,17 +219,52 @@ func (adm *Admin) listStorageNodes(ctx context.Context) ([]vmspb.StorageNodeMeta
 		return nil, status.Errorf(codes.Unavailable, "list storage nodes: cluster metadata not fetched")
 	}
 
+	lazyInit := false
+	lazyReplicasMap := make(map[types.StorageNodeID][]snpb.LogStreamReplicaMetadataDescriptor, len(md.StorageNodes))
+	getLazyReplicasMap := func() (map[types.StorageNodeID][]snpb.LogStreamReplicaMetadataDescriptor, error) {
+		if !lazyInit {
+			for _, lsd := range md.LogStreams {
+				for _, rd := range lsd.Replicas {
+					snid := rd.StorageNodeID
+					snd := md.GetStorageNode(snid)
+					if snd == nil {
+						return nil, status.Errorf(codes.Internal, "list storage node: inconsistent cluster metadata")
+					}
+					lazyReplicasMap[snid] = append(lazyReplicasMap[snid],
+						snpb.LogStreamReplicaMetadataDescriptor{
+							LogStreamReplica: varlogpb.LogStreamReplica{
+								StorageNode: snd.StorageNode,
+								TopicLogStream: varlogpb.TopicLogStream{
+									TopicID:     lsd.TopicID,
+									LogStreamID: lsd.LogStreamID,
+								},
+							},
+							Path: rd.Path,
+						},
+					)
+				}
+			}
+			lazyInit = true
+		}
+		return lazyReplicasMap, nil
+	}
+
 	snms := make([]vmspb.StorageNodeMetadata, 0, len(md.StorageNodes))
 	snmsMap := adm.statRepository.ListStorageNodes()
 	for _, snd := range md.StorageNodes {
 		if snm, ok := snmsMap[snd.StorageNodeID]; ok {
 			snms = append(snms, *snm)
 			continue
 		}
+		replicasMap, err := getLazyReplicasMap()
+		if err != nil {
+			return nil, err
+		}
 		snms = append(snms, vmspb.StorageNodeMetadata{
 			StorageNodeMetadataDescriptor: snpb.StorageNodeMetadataDescriptor{
-				ClusterID:   adm.cid,
-				StorageNode: snd.StorageNode,
+				ClusterID:         adm.cid,
+				StorageNode:       snd.StorageNode,
+				LogStreamReplicas: replicasMap[snd.StorageNodeID],
```

Comment:
What is this metadata intended for?
If `StorageNodeMetadata.LastHeartbeatTime` is zero, kovadm ignores the metadata.
Even if it is not ignored, `LogStreamReplicaMetadataDescriptor` that cannot be used by kovadm because `UpdateTime` is 0.

Reply:
Yes, you are right. It is not intended for the clients like KOV (Kafka On Varlog) admin.

As I said on messenger, its purpose is that restarted storage node checks mandatory directories that store log stream data. If there are no directories for log streams that are registered to the metadata repository, the storage node must not be started.